sentry/src/main/java/io/sentry/util/ExceptionUtils.java

```diff
@@ -0,0 +1,23 @@
+package io.sentry.util;
+
+import org.jetbrains.annotations.ApiStatus;
+import org.jetbrains.annotations.NotNull;
+
+@ApiStatus.Internal
+public final class ExceptionUtils {
+
+  /**
+   * Returns exception root cause or the exception itself if there are no causes
+   *
+   * @param throwable - the throwable
+   * @return the root cause
+   */
+  public static @NotNull Throwable findRootCause(final @NotNull Throwable throwable) {
+    Objects.requireNonNull(throwable, "throwable cannot be null");
+    Throwable rootCause = throwable;
+    while (rootCause.getCause() != null && rootCause.getCause() != rootCause) {
+      rootCause = rootCause.getCause();
+    }
+    return rootCause;
+  }
```

Comment:
while this would fix the problem, it'd also be a breaking change and introduce new issues.

See the use case:

`NotFoundException(IOException)`
`InternalError(IOException)`

both would be linked as `IOException` but either the App. wrapped into a known type or 3rd party libs did it, not sure if it's a good idea to do it here, but rather doing the check before setting the Span-Exception to the Hub WeakRef Map.

@bruno-garcia wdyt?

Reply:
What do you mean by "linked as"? Sentry events stay the same, its only about associating them with spans. I don't see how this is a breaking change

Reply:
there's an association between Span and Throwable right.
we're changing the behavior now, from the Throwable itself to its cause, given the example above with a dummy `NotFoundException` and `InternalError`.

Maybe a better solution would be either, getting the cause in such integrations (eg Spring AMQP and integrations that always wrap up in a generic Throwable) and passing to `throwableToSpan` the correct Throwable, the cause.

2nd, maybe a bit more expensive, which is setting to `throwableToSpan` every `Throwable` and its cause.

eg:
The example: NotFoundException(IOException)
`throwableToSpan` will contain 2 entries, with the `NotFoundException` and `IOException`

Reply:
lets see what @bruno-garcia says

Reply:
I am still not getting it. Can you show an example test that would be passing now, but fails after this change?

Reply:
tbh I am not sure how to explain differently, maybe we jump on a call? DM me

Reply:
Good idea to discuss this on a call. Lets sync up if y'all didn't yet

Reply:
Manoel and I talked btw on a call. We realized that we need to consider the root also when adding to the weak ref, but the type of the exception doesn't play a role so I think we can mark this one as resolved (I added a note for the other point)